### ISSUE TITLE
关于如何解读paddlenlp中文模型预测结果的问题

### ISSUE THREAD

**Issue by FleetCommander004** [PaddlePaddle/PaddleNLP#1196]
最近我在训练并使用paddlenlp的skep_ernie_1.0_large_ch模型。**我的paddlenlp版本是2.1.0，paddlepaddle-gpu版本是2.1.3.**
我发现最终模型预测出来的结果是这个格式的：
[[-3.4302268 ,  7.814872  , -3.0930223 ],
       [-3.8110301 ,  6.4573817 , -2.0783448 ],
       。。。
       [-2.7968476 ,  6.406624  , -2.7197244 ],
       [-3.7131455 ,  6.922563  , -2.263939  ]]
**我做的是一个三分类的模型，现在我在试图把这些数字转化为每一类的对应的概率。**
我用的部分代码如下：
[【参考https://aistudio.baidu.com/aistudio/projectdetail/2094427】]

**model = SkepForSequenceClassification.from_pretrained('skep_ernie_1.0_large_ch', num_classes=3)**
from paddle.static import InputSpec
input = InputSpec((-1, -1), dtype='int64', name='input')
model = paddle.Model(model, input)
model.load(checkpoint_path)

def open_func(file_path):
    return [line.strip() for line in open(file_path, 'r', encoding='utf8').readlines()[1:] if len(line.strip().split('\t')) >= 2]
train_raw = open_func("new_train1.tsv")
dev_raw = open_func("new_dev1.tsv")
test_raw = open_func("new_test1.tsv")
dev_q3_raw = open_func("new_dev1_q3.tsv")
test_q3_raw = open_func("new_test1_q3.tsv")
print(train_raw[0:10])
print(dev_raw[0:10])
#查看数据集的数据种类
print(type(train_raw))
#打开分词器
tokenizer = SkepTokenizer.from_pretrained('skep_ernie_1.0_large_ch')
#数据集正式转化为模型可分析的模式
from paddle.io import Dataset, DataLoader
from paddlenlp.data import Pad, Stack, Tuple
import numpy as np
label_list = [0, 1,2]

class MyDataset(Dataset):
    def __init__(self, data, tokenizer, max_len=512, for_test=False):
        super().__init__()
        self._data = data
        self._tokenizer = tokenizer
        self._max_len = max_len
        self._for_test = for_test
    
    def __len__(self):
        return len(self._data)
    
    def __getitem__(self, idx):
        samples = self._data[idx].split('\t')
        label = samples[-2]
        text = samples[-1]
        label = int(label)
        text = self._tokenizer.encode(text, max_seq_len=self._max_len)['input_ids']
        if self._for_test:
            return np.array(text, dtype='int64')
        else:
            return np.array(text, dtype='int64'), np.array(label, dtype='int64')
        
def batchify_fn(for_test=False):
    if for_test:
        return lambda samples, fn=Pad(axis=0, pad_val=tokenizer.pad_token_id): np.row_stack([data for data in fn(samples)])
    else:
        return lambda samples, fn=Tuple(Pad(axis=0, pad_val=tokenizer.pad_token_id),
                                        Stack()): [data for data in fn(samples)]
    
def get_data_loader(data, tokenizer, batch_size=32, max_len=512, for_test=False):
    dataset = MyDataset(data, tokenizer, max_len, for_test)
    shuffle = True if not for_test else False
    data_loader = DataLoader(dataset=dataset, batch_size=batch_size, collate_fn=batchify_fn(for_test), shuffle=shuffle)
    return data_loader

epochs = 5
learning_rate = 1e-5        # chnsenticorp 2e-5  /nlpcc14sc 2e-5
batch_size = 16    #chnsenticorp 64 / nlpcc14sc 128
max_len = 512
train_dataloader = get_data_loader(train_raw, tokenizer, batch_size, max_len, for_test=False)
dev_dataloader = get_data_loader(dev_raw, tokenizer, batch_size, max_len, for_test=False)
test_dataloader = get_data_loader(test_raw, tokenizer, batch_size, max_len, for_test=True)
dev_q3_dataloader = get_data_loader(dev_q3_raw, tokenizer, batch_size, max_len, for_test=False)
test_q3_dataloader = get_data_loader(test_q3_raw, tokenizer, batch_size, max_len, for_test=True)
#正式存储运输结果
predicts = []
for batch in test_dataloader:
    predict = model.predict_batch(batch)
    predicts += predict[0].argmax(axis=-1).tolist()
    
print(predicts)

import pandas as pd
data_dev = pd.read_csv("new_dev1.tsv",sep="\t",encoding="utf-8")
artificial_score = []
for index in range(0,len(predicts)):
    artificial_score.append(int(data_dev.iloc[index,1]))
from sklearn.metrics import confusion_matrix
print("new paddle model scores:")
conf_mx = confusion_matrix(artificial_score,predicts)
print(conf_mx)

predicts_q3 = []
predicts_list = []
for batch in test_q3_dataloader:
    predict = model.predict_batch(batch)
    predicts_list.append(predict)
    predicts_q3 += predict[0].argmax(axis=-1).tolist()

predicts_q3 = []
predicts_list = []
for batch in test_q3_dataloader:
    predict = model.predict_batch(batch)
    predicts_list.append(predict)
    predicts_q3 += predict[0].argmax(axis=-1).tolist()

**但是我试着用了logit function和sigmoid function来解，发现结果三个概率加起来大于一。**
然后我又发现源代码的SkepForSequenceClassification类别里面有logits这个内容。
-----
#来自PaddleNLP/paddlenlp/transformers/skep/modeling.py ，427-429行。
pooled_output = self.dropout(pooled_output)
        logits = self.classifier(pooled_output)
        return logits
-----

所以这个logitss是跟logistic function紧密相关的内容吗？我又应该如何把预测出来的一组组三个数字的numpy array转化为每一类的对应的概率呢？

**Comment by gongel** [comment]
你好，可以加一个`paddle.nn.functional.softmax`进行归一化。

**Comment by FleetCommander004** [comment]
> 你好，可以加一个`paddle.nn.functional.softmax`进行归一化。

您好，非常感谢您的回复，我会尽快尝试的。

**Comment by FleetCommander004** [comment]
> 你好，可以加一个`paddle.nn.functional.softmax`进行归一化。

试验过了，效果很好。非常感谢。

**Comment by chenxiaozeng** [comment]
Issue resolved. If you have other questions, please feel free to reopen it.
